### DISCUSSION TITLE
Getting some error when try to apply average zero condition on a scalar variable

### DISCUSSION THREAD

**Discussion by A123-bansal123** [firedrakeproject/firedrake#2901] [Firedrake support]
```
W = FunctionSpace(M, "Lagrange", 1)
p_exact = Function(W)
pressure=-0.5*(1+((1-2*x*x+2*y*y)/(r2*r2)))
p_exact.interpolate(pressure)

pressuredx = assemble(interpolate(p_exact, W) * dx)
size_domain = assemble(interpolate(Constant(1.0), W) * dx)
pressure = pressure - (pressuredx / size_domain)
p_exact.interpolate(pressure)
```
when I try to make p_exact belongs to L^2_0 i.e. p_exact average zero condition (int_\Omega p dx =0) then I get the error message 
```
/home/aparna/software/firedrake/src/tsfc/tsfc/loopy.py:499: RuntimeWarning: invalid value encountered in scalar subtract
  if r and numpy.abs(v - r) < ctx.epsilon:
```
Please help !

**Comment by dham** [comment]
It's a bit hard to test what you've done because your example is incomplete. `M`, `x`, `y`, and `r2` are undefined. If I make minimal changes to define all the symbols then the code executes without error:
```python3
from firedrake import *
M = UnitSquareMesh(2, 2)
W = FunctionSpace(M, "Lagrange", 1)
p_exact = Function(W)
x, y = SpatialCoordinate(M)
r2 =2.
pressure=-0.5*(1+((1-2*x*x+2*y*y)/(r2*r2)))
p_exact.interpolate(pressure)

pressuredx = assemble(interpolate(p_exact, W) * dx)
size_domain = assemble(interpolate(Constant(1.0), W) * dx)
pressure = pressure - (pressuredx / size_domain)
p_exact.interpolate(pressure)
```

I should also point out that two of your interpolations are unnecessary. The following code would do the same as your code:

```python3
from firedrake import *
M = UnitSquareMesh(2, 2)
W = FunctionSpace(M, "Lagrange", 1)
p_exact = Function(W)
x, y = SpatialCoordinate(M)
r2 = 2.
pressure=-0.5*(1+((1-2*x*x+2*y*y)/(r2*r2)))
p_exact.interpolate(pressure)

pressuredx = assemble(p_exact * dx)
size_domain = assemble(Constant(1.0, M) * dx)
pressure = pressure - (pressuredx / size_domain)
p_exact.interpolate(pressure)
```

**Comment by A123-bansal123** [comment]
sir, it still shows same error 
```
/home/aparna/software/firedrake/src/tsfc/tsfc/loopy.py:499: RuntimeWarning: invalid value encountered in scalar subtract
  if r and numpy.abs(v - r) < ctx.epsilon:
```
I am using  a mesh with a hole. 
i am using a 3 D mesh. May be that's why it shows an error.

**Reply by dham** [reply]
Are you saying that exactly the code that I posted creates an error, or that your code, which is different from my code, produces the error? I think you are saying the latter. In which case please post a complete minimal failing example, including all the code I need to recreate the error, and the mesh file.

**Comment by A123-bansal123** [comment]
Your  code works well. My code is showing error when I impose this average
zero condition and I also made the changes that you mentioned. That's why I
think, may be the error occur due to 3d mesh.

**Reply by dham** [reply]
At this stage, you haven't shown your code, so it's impossible to say. What you need to do, as I note above, is post a complete minimal failing example. By this, I mean a complete program, in which all the variables are defined, that exhibits the problem, along with any data needed to run it. It would be easiest if you could use one of the in-built utility meshes (e.g. `UnitCubeMesh`), but if you need to post a mesh file then that is also OK.

**Comment by A123-bansal123** [comment]
Thank you Now it works well.
